ast.go

```diff
@@ -26,20 +26,101 @@ import (
 // The Serializer interface is implemented by all
 // expressions/statements.
 type Serializer interface {
-	// Serialize writes the statement/expression to the io.Writer. If an
-	// error is returned the io.Writer may contain partial output.
-	Serialize(w io.Writer) error
+	// Serialize writes the statement/expression to the sqlWriter. If an
+	// error is returned the sqlWriter may contain partial output.
+	Serialize(w sqlWriter) error
 }
 
 // Serialize serializes a serializer to a string.
 func Serialize(s Serializer) (string, error) {
 	var buf bytes.Buffer
-	if err := s.Serialize(&buf); err != nil {
+	if err := s.Serialize(&standardWriter{&buf}); err != nil {
```

Comment:
Might be better to embed the `bytes.Buffer` directly in `standardWriter` and `placeholderWriter` as doing so avoids an allocation. Since those types are not exported the flexibility of being able to take any `io.Writer` isn't warranted.